modules/javafx.web/src/main/java/com/sun/webkit/network/HTTP2Loader.java

```diff
@@ -0,0 +1,648 @@
+                } else {
+                    this.subscription = subscription;
+                    requestIfNotCancelled();
+                }
+            }
+
+            private void requestIfNotCancelled() {
+                if (canceled) {
+                    subscription.cancel();
+                } else {
+                    subscription.request(1);
+                }
+            }
+        });
+        return normalBodySubscriber;
+    }
+
+    private BodySubscriber<Void> getBodySubscriber(final String contentEncoding) {
+        return contentEncoding.isEmpty() ?
+                  createNormalBodySubscriber() : createZIPEncodedBodySubscriber(contentEncoding);
+    }
+
+    private HTTP2Loader(WebPage webPage,
+              ByteBufferPool byteBufferPool,
+              boolean asynchronous,
+              String url,
+              String method,
+              String headers,
+              FormDataElement[] formDataElements,
+              long data)
+    {
+        this.webPage = webPage;
+        this.asynchronous = asynchronous;
+        this.url = url;
+        this.method = method;
+        this.headers = headers;
+        this.formDataElements = formDataElements;
+        this.data = data;
+
+        URI uri;
+        try {
+            uri = toURI();
+        } catch(MalformedURLException e) {
+            this.response = null;
+            didFail(e);
+            return;
+        }
+
+        final var request = HttpRequest.newBuilder()
+                               .uri(uri)
+                               .headers(getRequestHeaders()) // headers from WebCore
+                               .headers(getCustomHeaders()) // headers set by us
+                               .version(Version.HTTP_2)  // this is the default
+                               .method(method, getFormDataPublisher())
+                               .build();
+
+        final BodyHandler<Void> bodyHandler = rsp -> {
+            if(!handleRedirectionIfNeeded(rsp)) {
+                didReceiveResponse(rsp);
+            }
+            return getBodySubscriber(getContentEncoding(rsp));
+        };
+
+        // Run the HttpClient in the page's access control context
+        this.response = AccessController.doPrivileged((PrivilegedAction<CompletableFuture<Void>>) () -> {
+            return HTTP_CLIENT.sendAsync(request, bodyHandler)
+                              .thenAccept($ -> {})
+                              .exceptionally(ex -> didFail(ex.getCause()));
+        }, webPage.getAccessControlContext());
+
+        if (!asynchronous) {
+            waitForRequestToComplete();
+        }
+    }
+
+    /**
+     * Cancels this loader.
+     */
+    @Override
+    public void fwkCancel() {
+        if (logger.isLoggable(Level.FINEST)) {
+            logger.finest(String.format("data: [0x%016X]", data));
+        }
+        canceled = true;
+    }
+
+    private void callBackIfNotCancelled(final Runnable r) {
```

Comment:
Minor nit: could be renamed to `callBackIfNotCanceled` -- with one `l` -- for consistency.